bindings/sundials4py/test/test_idas.py

```diff
@@ -0,0 +1,225 @@
+    assert status == IDA_SUCCESS
+
+    # Sensitivity setup
+    Ns = 1  # number of sensitivities (wrt yy0)
+    ism = IDA_SIMULTANEOUS
+    yyS0 = [N_VClone(yy) for _ in range(Ns)]
+    ypS0 = [N_VClone(yp) for _ in range(Ns)]
+    for v in yyS0:
+        N_VConst(0.0, v)
+    for v in ypS0:
+        N_VConst(0.0, v)
+
+    def resS(Ns, t, yy, yp, resval, yS, ypS, resvalS, _, tmp1, tmp2, tmp3):
+        # Sensitivity residuals: d(res)/d(yy) * yyS + d(res)/d(yp) * ypS
+        # For smoke test, just zero out
+        for i in range(Ns):
+            N_VConst(0.0, resvalS[i])
+        return 0
+
+    status = IDASensInit(ida.get(), Ns, ism, resS, yyS0, ypS0)
+    assert status == IDA_SUCCESS
+
+    status = IDASensSStolerances(ida.get(), 1e-4, np.array([1e-4], dtype=sunrealtype))
+    assert status == IDA_SUCCESS
+
+    tout = ode_problem.TF
+    yySout = [N_VClone(yy) for _ in range(Ns)]
+    ypSout = [N_VClone(yp) for _ in range(Ns)]
+    status, tret = IDASolve(ida.get(), tout, yy, yp, IDA_NORMAL)
+    assert status == IDA_SUCCESS
+
+    # Get sensitivities (smoke test: just check call and shape)
+    status, tret_sens = IDAGetSens(ida.get(), yySout)
+    assert status == IDA_SUCCESS
+    assert len(yySout) == Ns
+
+
+def test_idas_adjoint(sunctx):
+    ode_problem = AnalyticDAE()
+
+    ida = IDACreate(sunctx)
+    yy = N_VNew_Serial(2, sunctx)
+    yp = N_VNew_Serial(2, sunctx)
+
+    # y and y' initial conditions
+    ode_problem.set_init_cond(yy, yp, ode_problem.T0)
+
+    ls = SUNLinSol_SPGMR(yy, SUN_PREC_LEFT, 0, sunctx)
+
+    def resfn(t, yy, yp, rr, _):
+        return ode_problem.res(t, yy, yp, rr)
+
+    def psolve(t, yy, yp, rr, r, z, cj, delta, _):
+        return ode_problem.psolve(t, yy, yp, rr, r, z, cj, delta)
+
+    status = IDAInit(ida.get(), resfn, 0.0, yy, yp)
+    assert status == IDA_SUCCESS
+
+    status = IDASStolerances(ida.get(), 1e-4, 1e-4)
+    assert status == IDA_SUCCESS
+
+    status = IDASetLinearSolver(ida.get(), ls, None)
+    assert status == IDA_SUCCESS
+
+    status = IDASetPreconditioner(ida.get(), None, psolve)
+    assert status == IDA_SUCCESS
+
+    # Adjoint (backward) problem setup
+    steps = 5
+    interp = IDA_HERMITE
+    status = IDAAdjInit(ida.get(), steps, interp)
+    assert status == IDA_SUCCESS
+
+    # Integrate forward
+    tout = ode_problem.TF
+    status, tret, ncheck = IDASolveF(ida.get(), tout, yy, yp, IDA_NORMAL)
+    assert status == IDA_SUCCESS
+
+    # Create backward problem
+    yyB = N_VClone(yy)
+    ypB = N_VClone(yp)
+    N_VConst(0.0, yyB)
+    N_VConst(0.0, ypB)
+    lsB = SUNLinSol_SPGMR(yyB, 0, 0, sunctx)
+
+    status, whichB = IDACreateB(ida.get())
+    assert status == IDA_SUCCESS
+
+    def resB(t, y, yp, yB, ypB, resvalB, _):
+        N_VConst(0.0, resvalB)
+        return 0
+
+    status = IDAInitB(ida.get(), whichB, resB, tout, yyB, ypB)
+    assert status == IDA_SUCCESS
+
+    status = IDASStolerancesB(ida.get(), whichB, 1e-4, 1e-4)
```

Comment:
Use `SUNREALTYPE_RTOL` and `SUNREALTYPE_ATOL`?
```suggestion
    status = IDASStolerancesB(ida.get(), whichB, SUNREALTYPE_RTOL, SUNREALTYPE_ATOL)
```